T24004 Comprehensive unit tests

pwithnall opened the pull request:
https://phabricator.endlessm.com/T24004

pwithnall committed:
tests: Add comprehensive tests for libeos-parental-controls

These test get/set behaviour of EpcAppFilter, GAppInfo support, and
various error handling paths and corner cases.

With these tests, coverage of app-filter.c is:
 • Lines: 97%
 • Functions: 100%
 • Branches: 57.5%

The branch coverage is not correctly reported, as it factors in the
failure branches for g_return_if_fail() precondition assertions, which
we deliberately do not test. With those branches manually excluded,
branch coverage is around 89% instead.

This adds libglib-testing as a subproject dependency, for its
GtDBusQueue. It adds gio-unix-2.0 as a dependency of the app-filter
tests, in order to be able to construct GDesktopAppInfos.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T24004

pwithnall committed:
lib: Fix handling of accountsservice queries for missing users

Since we ported to using FindUserById, it returns an explicit Failed
error if a user doesn’t exist. Previously, we would guess at the user’s
object path and call a method on it, and would receive a D-Bus ‘method
not found’ error in response if the user didn’t exist.

Correctly handle the explicit error from FindUserById.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T24004

pwithnall committed:
lib: Fix GVariant text format for default property value

It was a string in GVariant text format, not a GVariant format string,
so should have been passed to the parse function.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T24004

pwithnall committed:
lib: Fix handling of X-Flatpak-RenamedFrom desktop file keys

A variable was being initialised self-referentially, which broke
handling of the X-Flatpak-RenamedFrom key.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T24004

pwithnall committed:
lib: Fix support for custom GDBusConnections when setting filters

Previously, a caller-provided custom GDBusConnection was dropped on the
ground.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T24004

pwithnall committed:
accounts-service: Rename D-Bus properties to CamelCase

That’s what’s more conventional for D-Bus properties, and we really
should have used CamelCase from the beginning.

See the advice on
https://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-properties:
> Strictly speaking, D-Bus property names are not required to follow
> the same naming restrictions as member names, but D-Bus property
> names that would not be valid member names (in particular,
> GObject-style dash-separated property names) can cause
> interoperability problems and should be avoided.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

GeorgesStavracas commented on libeos-parental-controls/tests/app-filter.c:
```diff
@@ -236,6 +289,972 @@ test_app_filter_builder_copy_full (void)
+    };
+
+  gt_dbus_queue_set_server_func (fixture->queue, get_app_filter_server_cb,
+                                 &get_app_filter_data);
```
This discards `const`

GeorgesStavracas commented on libeos-parental-controls/tests/app-filter.c:
```diff
@@ -236,6 +289,972 @@ test_app_filter_builder_copy_full (void)
+    };
+
+  gt_dbus_queue_set_server_func (fixture->queue, get_app_filter_server_cb,
+                                 &get_app_filter_data);
```
This too.

GeorgesStavracas commented on libeos-parental-controls/tests/app-filter.c:
```diff
@@ -236,6 +289,972 @@ test_app_filter_builder_copy_full (void)
+    };
+
+  gt_dbus_queue_set_server_func (fixture->queue, get_app_filter_server_cb,
+                                 &get_app_filter_data);
```
And this

GeorgesStavracas commented on libeos-parental-controls/tests/app-filter.c:
```diff
@@ -236,6 +289,972 @@ test_app_filter_builder_copy_full (void)
+  g_autofree const gchar **oars_sections = NULL;
+
+  gt_dbus_queue_set_server_func (fixture->queue, get_app_filter_server_cb,
+                                 &get_app_filter_data);
```
... here too

GeorgesStavracas reviewed (approved):
These tests look good (very good, in fact), but there are 4 small compiler warnings being triggered.

pwithnall commented:
> there are 4 small compiler warnings being triggered.

What warning flags are you using? I don’t get those warnings locally, and Jenkins doesn’t get them either. It would be good to enable such a warning in the top-level `meson.build`.

GeorgesStavracas commented:
`-Wdiscarded-qualifiers` I guess. Notice that Meson outputs:

`Compiler for C supports arguments -Wno-discarded-qualifiers -Wdiscarded-qualifiers: YES`

So we're checking for this argument, but not using it.

Edit: and "by not using it", I mean that this specific argument is not present at the Meson-generated `compile_commands.json` file.

wjt commented on libeos-parental-controls/tests/app-filter.c:
```diff
@@ -236,6 +289,972 @@ test_app_filter_builder_copy_full (void)
+      gboolean expected_allowed;
+      const gchar *key_file_data;
+    }
+  vectors[] =
```
As an aside, I tend to do this kind of thing with one call to `g_test_add_data_func` per test data, perhaps because I am spoiled by [`@pytest.mark.parametrize`](https://docs.pytest.org/en/latest/parametrize.html) / [JUnit's less-pleasing equivalent](https://github.com/junit-team/junit4/wiki/Parameterized-tests). The nice thing about that approach is (of course) that if one case fails, the rest still run; it's also easier to see from the test name (either an integer index, or an explicit name on each test) which one is failing. If one of these tests fails in CI, the test name + assertion message will give no hint as to which it is.

pwithnall replied:
I would like to do that, but the downsides are:
 1. It moves the test data away from the test code.
 2. The combination of `struct`s not being valid inline values and declarations needing to be at the start of the function means that `main()` becomes an incomprehensible mess.

On balance, I think those downsides are reason to use the approach I’ve used here, but I completely acknowledge the cons you give, and wish there was some way around them.

Maybe a `g_test_start_subtest()` call which could be used inside a test, a bit like `g_test_skip()`? That would fix your second con, but can’t fix the first. I’ve kind of done that already with the `g_test_message()` call in the loop, though.

Grumble, nothing is satisfying. :-(

wjt replied:
Well, you could move the definition to the top level and leave it defined here, just reference it from main(). But fair enough.

wjt commented on libeos-parental-controls/tests/app-filter.c:
```diff
@@ -236,6 +289,972 @@ test_app_filter_builder_copy_full (void)
+  g_assert_cmpint (user_id, ==, data->expected_uid);
+
+  object_path = g_strdup_printf ("/org/freedesktop/Accounts/User%u", (uid_t) user_id);
+  reply1 = g_strdup_printf ("(@o '%s',)", object_path);
```
This and the following line seem like a weird way to spell `g_variant_new ("(o)", object path)`

pwithnall replied:
You are very right.

pwithnall replied:
Fixup pushed.

wjt commented on libeos-parental-controls/tests/app-filter.c:
```diff
@@ -236,6 +289,972 @@ test_app_filter_builder_copy_full (void)
+  /* Handle the FindUserById() call. */
+  gint64 user_id;
+  invocation1 =
+      gt_dbus_queue_assert_pop_message (queue,
```
Using this API turns out pretty nicely. (It really does remind me of the Telepathy testing framework, [eg](https://github.com/TelepathyIM/telepathy-gabble/blob/master/tests/twisted/text/receipts.py#L30). Probably not a coincidence.)

wjt commented on libeos-parental-controls/app-filter.c:
```diff
@@ -303,7 +303,7 @@ epc_app_filter_is_appinfo_allowed (EpcAppFilter *filter,
 
           for (gsize i = 0; old_flatpak_apps[i] != NULL; i++)
             {
-              gchar *old_flatpak_app = g_strstrip (old_flatpak_app);
```
If I compile the following test program with `-Wall`:

```c
char *foo (char *x) { return x; }

void  main(void)
{
  char *bar = foo(bar);
}
```

I get a warning:

```
bar.c: In function ‘main’:
bar.c:5:15: warning: ‘bar’ is used uninitialized in this function [-Wuninitialized]
   char *bar = foo(bar);
               ^~~~~~~~
```

More zealous warning level needed?

pwithnall replied:
Hmm, interesting. Putting that code into `app-filter.c` gives me a warning. Undoing the `old_flatpak_app` fix doesn’t give me a warning. Putting the following code (i.e. wrapping your code in a loop) into `app-filter.c` also *doesn’t* give me a warning:
```
  for (gsize i = 0; i < 10; i++)
    {
      char *bar = g_strstrip (bar);
    }
```

So it looks like the compiler’s heuristics for detecting `-Wmaybe-uninitialized` get broken by loops.

I’m using GCC 8.2.1. The bug might be https://gcc.gnu.org/bugzilla/show_bug.cgi?id=54554, since I’m compiling at `-O0`; but looking at the list of ‘uninitialized’ bugs in the GCC Bugzilla, there’s enough there that it’s either been reported already, or reporting it would not get it fixed.

In any case, Meson does seem to be successfully passing the warning flags in to the compiler command.

wjt commented on libeos-parental-controls/tests/app-filter.c:
```diff
@@ -43,6 +46,26 @@ assert_strv_equal (const gchar * const *strv_a,
 }
 
+/* FIXME: Use g_assert_cmpvariant() when
+ * https://gitlab.gnome.org/GNOME/glib/issues/1191 is fixed. */
```
Nice, I've definitely open-coded this in at least 2 projects before.

pwithnall committed:
build: Enable -Wdiscarded-qualifiers

The codebase is clean for it, so why not enforce it?

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T24004

pwithnall commented:
Fixup pushed, thanks. The fixup will need to be squashed before this can be merged. This branch also still depends on https://gitlab.gnome.org/pwithnall/libglib-testing/merge_requests/4 being reviewed and merged in libglib-testing upstream.

wjt commented on libeos-parental-controls/tests/app-filter.c:
```diff
@@ -236,6 +289,972 @@ test_app_filter_builder_copy_full (void)
+      g_assert_cmpstr (property_interface, ==, "com.endlessm.ParentalControls.AppFilter");
+      g_assert_cmpstr (property_name, ==, expected_properties[i]);
+
+      if (data->error_index >= 0 && (gsize) data->error_index == i)
```
I feel like making `i` have the same type as `data->error_index` would be neater than needing this cast. (Do you even need this cast to a wider type, actually?)

I know that conceptually, an array index should be of type `size_t`, but in this case, we know its largest value is `4`.

pwithnall replied:
Doing it this way means that the compiler forces you to handle the cases where `data->error_index < 0`, which is good, because we needed to handle them explicitly (i.e. they are possible and callers do set `error_index` to negative values).

wjt replied:
Fair enough.

wjt commented on libeos-parental-controls/tests/app-filter.c:
```diff
@@ -236,6 +289,972 @@ test_app_filter_builder_copy_full (void)
+/* Test that epc_set_app_filter() returns an appropriate error if the mock D-Bus
+ * service reports that the given user cannot be found.
+ *
+ * The mock D-Bus replies are generated inline. */
```
What's the rationale for doing some of these tests in the "worker thread" style, and some in this "async + inline replies" style? I guess it does demonstrate both models of GtDBusQueue usage.

pwithnall replied:
The worker thread style is needed whenever a test uses the synchronous libeos-parental-controls API. Any other test which can sensibly reuse the same worker thread function also uses it. Everything else is done in the ‘inline replies’ style because it keeps the client/server conversation flow in a single function, so it should be a bit more obvious what’s happening.

And yes, it is partially to demonstrate both models of `GtDBusQueue` usage.

Does that make sense?

wjt replied:
Yep!

pwithnall committed:
lib: Simplify creation of a GVariant

This introduces no functional changes.

Signed-off-by: Philip Withnall <withnall@endlessm.com>

https://phabricator.endlessm.com/T24004

GeorgesStavracas commented:
Hm, this branch breaks the GNOME Settings panel. The app filter is being denied:

`(gnome-control-center:20425): user-accounts-cc-panel-WARNING **: Error retrieving app filter for user 'georges': Not allowed to query app filter data for user 1001`

pwithnall commented:
> Hm, this branch breaks the GNOME Settings panel. The app filter is being denied:
> 
> `(gnome-control-center:20425): user-accounts-cc-panel-WARNING **: Error retrieving app filter for user 'georges': Not allowed to query app filter data for user 1001`

Have you installed the updated `accounts-service/com.endlessm.ParentalControls.AppFilter.xml` file from eos-parental-controls and restarted `accounts-daemon.service` since then? This is likely being caused by the renaming of the properties — if accountsservice is returning the old property names to libeos-parental-controls, if will assume the lack of new properties means that permission was denied to access them.

GeorgesStavracas commented:
> Have you installed the updated accounts-service/com.endlessm.ParentalControls.AppFilter.xml file from eos-parental-controls and restarted accounts-daemon.service since then? This is likely being caused by the renaming of the properties — if accountsservice is returning the old property names to libeos-parental-controls, if will assume the lack of new properties means that permission was denied to access them.

That, and a corrupted .xml file (missing the last couple lines) were breaking everything. Tested on a fresh install, with a correct file, restarted the accountsservice daemon, and everything's working as it should.

GeorgesStavracas commented:
Please squash the fixups so it can be merged

pwithnall commented:
Squashed and rebased on master. I’ll wait for the tests to pass then merge. Thanks.
